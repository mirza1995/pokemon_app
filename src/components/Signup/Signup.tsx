import { useReducer, useState } from 'react';
import { FormattedMessage } from 'react-intl';

import usersService from '@services/usersService';
import { LoginSignupResponse } from '@models/LoginSignupResponse';
import { User } from '@models/User';
import { UserError } from '@models/UserError';
import { CreateUserDto } from '@models/CreateUserDto';
import { formatDate } from '@utils/date';
import { isValidEmail, isValidPassword } from '@utils/validation';
import { PrimaryButton } from '@components/PrimaryButton';
import { DatePicker } from '@components/DatePicker';
import { Input } from '@components/Input';

import { ACTIONS } from '@reducers/actions';
import { userErrorReducer } from '@reducers/UserErrorReducer';
import { userReducer } from '@reducers/UserReducer';

type SignupProps = {
  onSuccessfulSignup: () => void;
};

export const Signup = ({ onSuccessfulSignup }: SignupProps) => {
  const [user, dispatchUser] = useReducer(userReducer, {});
  const [errors, dispatchError] = useReducer(userErrorReducer, {});
  const [signupError, setSignupError] = useState<string>('');

  const onValueChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value, name } = e.currentTarget;
    dispatchError({ type: ACTIONS.REMOVE, data: {}, errorToRemove: name });
    dispatchUser({ type: ACTIONS.UPDATE, data: { [name]: value } });
  };

  const onBirthDateChange = (newBirthDate: Date) => {
    dispatchUser({ type: ACTIONS.UPDATE, data: { birthDate: newBirthDate } });
  };

  const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    handleSignup();
  };

  const handleSignup = async () => {
    setSignupError('');
    if (validate()) {
      const response: LoginSignupResponse = await usersService.createUser(
        getCreateUserDto(user as User)
      );
      if (response.error) {
        setSignupError(response.error);
      } else {
        onSuccessfulSignup();
      }
    }
  };

  const validate = () => {
    const errors: Partial<UserError> = {};

    if (!isValidEmail(user.email)) {
      errors.email = 'errors.email.invalid';
    }

    if (!isValidPassword(user.password)) {
      errors.password = 'errors.password.invalid';
    }

    if (!user.firstName?.length) {
      errors.firstName = 'errors.firstName.invalid';
    }

    if (!user.lastName?.length) {
      errors.lastName = 'errors.lastName.invalid';
    }

    dispatchError({ type: ACTIONS.INIT, data: errors });

    return !Object.keys(errors).length;
  };

  const getCreateUserDto = (user: User): CreateUserDto => {
    return {
      email: user.email,
      password: user.password,
      first_name: user.firstName,
      last_name: user.lastName,
      date_of_birth: formatDate(user.birthDate)
    };
  };

  return (
    <div className='modal-default md:w-[440px]'>
      <h4 className='text-center text-lg font-medium'>
        <FormattedMessage id='signup.create' />
      </h4>

      <form onSubmit={onSubmit} className='mt-5'>
        <div>
          <div className='flex gap-[10px]'>
            <Input
              name='firstName'
              value={user.firstName || ''}
              type='text'
              labelId='signup.firstName'
              errorId={errors.firstName}
              onValueChange={onValueChange}
            />
            <Input
              name='lastName'
              value={user.lastName || ''}
              type='text'
              labelId='signup.lastName'
              errorId={errors.lastName}
              onValueChange={onValueChange}
            />
          </div>

          <DatePicker
            name='birthDate'
            labelId='signup.birthDate'
            selectedDate={user.birthDate || new Date('12.31.2015')}
            onSelect={onBirthDateChange}
          />

          <Input
            name='email'
            value={user.email || ''}
            type='text'
            labelId='signup.email'
            errorId={errors.email}
            onValueChange={onValueChange}
          />

          <Input
            name='password'
            value={user.password || ''}
            type='password'
            labelId='signup.password'
            errorId={errors.password}
            onValueChange={onValueChange}
          />
        </div>

        {signupError && (
          <div className='mt-5 text-md text-red-500'>{signupError}</div>
        )}

        <div className='mt-5'>
          <PrimaryButton type='submit' textId='signup.button' />
        </div>
      </form>
    </div>
  );
};
