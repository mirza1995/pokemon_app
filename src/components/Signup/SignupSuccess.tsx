import { FormattedMessage } from 'react-intl';
import { PrimaryButton } from '@components/PrimaryButton';

type SignupSuccessProps = {
  onOkClick: () => void;
};

export const SignupSuccess = ({ onOkClick }: SignupSuccessProps) => {
  return (
    <div className='modal-default'>
      <h4 className='text-center text-lg font-medium'>
        <FormattedMessage id='signup.success' />
      </h4>

      <div className='mx-auto mt-5 w-[150px]'>
        <PrimaryButton onClick={onOkClick} textId='ok' />
      </div>
    </div>
  );
};
