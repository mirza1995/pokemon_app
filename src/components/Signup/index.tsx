export * from './Signup';
export * from './SignupSuccess';
