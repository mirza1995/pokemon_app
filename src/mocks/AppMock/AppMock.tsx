import { Provider } from 'react-redux';
import { CookiesProvider } from 'react-cookie';
import { IntlProvider } from 'react-intl';
import English from 'lang/en.json';
import store from 'store/store';

//Mocks application to be used in tests
export const AppMock = ({ children }: { children: React.ReactNode }) => {
  return (
    <Provider store={store}>
      <CookiesProvider>
        <IntlProvider locale={navigator.language} messages={English}>
          {children}
        </IntlProvider>
      </CookiesProvider>
    </Provider>
  );
};
