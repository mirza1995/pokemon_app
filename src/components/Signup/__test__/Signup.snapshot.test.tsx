/* eslint-disable @typescript-eslint/no-empty-function */
import renderer from 'react-test-renderer';
import { AppMock } from '@mocks/AppMock';
import { Signup } from '@components/Signup';

it('Signup component renders correctly', () => {
  const tree = renderer
    .create(
      <AppMock>
        <Signup onSuccessfulSignup={() => {}} />
      </AppMock>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
