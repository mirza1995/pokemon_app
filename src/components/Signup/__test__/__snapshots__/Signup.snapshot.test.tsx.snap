// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component renders correctly 1`] = `
<div
  className="modal-default md:w-[440px]"
>
  <h4
    className="text-center text-lg font-medium"
  >
    Create an Account
  </h4>
  <form
    className="mt-5"
    onSubmit={[Function]}
  >
    <div>
      <div
        className="flex gap-[10px]"
      >
        <div
          className="relative mt-2.5 flex-1"
          onClick={[Function]}
        >
          <input
            className="
          block h-[50px] w-full rounded-3 border bg-gray-1 px-4
          pb-2.5 pt-6 text-md 
          font-normal text-primary-dark border-gray-2 
          peer appearance-none
          focus:border-primary focus:outline-none focus:ring-0"
            data-testid="firstName"
            name="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder=" "
            type="text"
            value=""
          />
          <label
            className=" absolute top-6 left-4 z-10 origin-[0] -translate-y-4 scale-75 text-md font-normal text-gray duration-300 peer-placeholder-shown:translate-y-0  peer-placeholder-shown:scale-100 peer-focus:-translate-y-4  peer-focus:scale-75"
          >
            First Name
          </label>
        </div>
        <div
          className="relative mt-2.5 flex-1"
          onClick={[Function]}
        >
          <input
            className="
          block h-[50px] w-full rounded-3 border bg-gray-1 px-4
          pb-2.5 pt-6 text-md 
          font-normal text-primary-dark border-gray-2 
          peer appearance-none
          focus:border-primary focus:outline-none focus:ring-0"
            data-testid="lastName"
            name="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder=" "
            type="text"
            value=""
          />
          <label
            className=" absolute top-6 left-4 z-10 origin-[0] -translate-y-4 scale-75 text-md font-normal text-gray duration-300 peer-placeholder-shown:translate-y-0  peer-placeholder-shown:scale-100 peer-focus:-translate-y-4  peer-focus:scale-75"
          >
            Last Name
          </label>
        </div>
      </div>
      <div>
        <div
          className="relative mt-2.5 flex-1"
          onClick={[Function]}
        >
          <input
            className="
          block h-[50px] w-full rounded-3 border bg-gray-1 px-4
          pb-2.5 pt-6 text-md 
          font-normal text-primary-dark border-gray-2 
          peer appearance-none
          focus:border-primary focus:outline-none focus:ring-0"
            data-testid="birthDate"
            name="birthDate"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            placeholder=" "
            type="text"
            value="December 31, 2015"
          />
          <label
            className=" absolute top-6 left-4 z-10 origin-[0] -translate-y-4 scale-75 text-md font-normal text-gray duration-300 peer-placeholder-shown:translate-y-0  peer-placeholder-shown:scale-100 peer-focus:-translate-y-4  peer-focus:scale-75"
          >
            Date of Birth
          </label>
        </div>
      </div>
      <div
        className="relative mt-2.5 flex-1"
        onClick={[Function]}
      >
        <input
          className="
          block h-[50px] w-full rounded-3 border bg-gray-1 px-4
          pb-2.5 pt-6 text-md 
          font-normal text-primary-dark border-gray-2 
          peer appearance-none
          focus:border-primary focus:outline-none focus:ring-0"
          data-testid="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          placeholder=" "
          type="text"
          value=""
        />
        <label
          className=" absolute top-6 left-4 z-10 origin-[0] -translate-y-4 scale-75 text-md font-normal text-gray duration-300 peer-placeholder-shown:translate-y-0  peer-placeholder-shown:scale-100 peer-focus:-translate-y-4  peer-focus:scale-75"
        >
          Email Address
        </label>
      </div>
      <div
        className="relative mt-2.5 flex-1"
        onClick={[Function]}
      >
        <input
          className="
          block h-[50px] w-full rounded-3 border bg-gray-1 px-4
          pb-2.5 pt-6 text-md 
          font-normal text-primary-dark border-gray-2 
          peer appearance-none
          focus:border-primary focus:outline-none focus:ring-0"
          data-testid="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          placeholder=" "
          type="password"
          value=""
        />
        <label
          className=" absolute top-6 left-4 z-10 origin-[0] -translate-y-4 scale-75 text-md font-normal text-gray duration-300 peer-placeholder-shown:translate-y-0  peer-placeholder-shown:scale-100 peer-focus:-translate-y-4  peer-focus:scale-75"
        >
          Password
        </label>
      </div>
    </div>
    <div
      className="mt-5"
    >
      <button
        className="w-full rounded-3 bg-primary-light py-4.5 text-base text-white shadow-3xl"
        onClick={[Function]}
        type="submit"
      >
        Create Account
      </button>
    </div>
  </form>
</div>
`;
